File: src/wordleSolver.java
//crane -> seats -> email
import java.io.*;
import java.util.*;
import java.util.List;
import java.util.Scanner;
import java.util.Iterator;


class wordleChecker {
    List<String> wordList = new ArrayList<>();
    List<String> badWordList = new ArrayList<>();
    public boolean running = true;

    HashMap<Character,Integer> dict = new HashMap<Character, Integer>();


    Character[] finWord = new Character[5];
    Scanner stdin = new Scanner(System.in);
    String word, result;
    List<Character> possibleLetters = new ArrayList<Character>();
    List<Character> badLetters = new ArrayList<Character>();

    List<Character> finalWord = new ArrayList<Character>();
    boolean allGreen = false;


    /**
     * Default constructor, sets instance variables to default values
     */
    public wordleChecker() {

    }

    public wordleChecker(List<String> list) {
        wordList = list;
    }

    public void initCheck() {
        System.out.print("What word was entered? (Common starter word is Crane) \n");
        word = stdin.next();

        System.out.println("Successfully initialized word: " + word);

        System.out.print("Enter the results of the initial guess in order from first letter to last letter (B = black (not used), G = green (correct spot), Y = yellow (used, but in the wrong spot), for example: BGYYB \n");
        result = stdin.next();
        letterSorter(word, result);
        //checker();

        wordRemover();
        System.out.println("wordRemover() finished");

        while (running){
            checker();
        }

    }

    public void checker() {

        System.out.print("What is the next word entered? \n");
        word = stdin.next();
        System.out.println("Successfully initialized word: " + word);
        System.out.print("Enter the results of the initial guess in order from first letter to last letter (B = black (not used), G = green (correct spot), Y = yellow (used, but in the wrong spot), for example: BGYYB \n");
        result = stdin.next();
        letterSorter(word, result);
        wordRemover();
        System.out.println("Is the word finished? Y/N");
        String yn = stdin.next();
        if (yn.equals("Y") || yn.equals("y")){
                running = false;
        }

    }


    public void printStatus() {
        System.out.println("The possible letters so far are: ");
        for (int i = 0; i <= possibleLetters.size() - 1; i++) {
            System.out.print(possibleLetters.get(i) + " ");
        }
        System.out.println("");

        System.out.println("The bad letters so far are: ");
        for (int i = 0; i <= badLetters.size() - 1; i++) {
            System.out.print(badLetters.get(i) + " ");

        }
        System.out.println("");

        System.out.println("The word we have so far is: ");
        for (int i = 0; i <= finWord.length - 1; i++) {
            System.out.print(finWord[i]);
        }
        System.out.println();
    }

    public void printPossibleWords() {
        for (int i = 0; i <= wordList.size() - 1; i++) {
            if (i % 10==0){
                System.out.print("\n");
            }
            System.out.print(wordList.get(i) + " ,");
        }
        System.out.println("There are " + wordList.size() + " words");
    }


    public void letterSorter(String input, String result) {
        for (int i = 0; i <= result.length() - 1; i++) {
            if (result.charAt(i) == 'B' || result.charAt(i) == 'b') { //checks if letter is black
                if (badLetters.contains(input.charAt(i)) || possibleLetters.contains(input.charAt(i))){
                    System.out.println("Letter " + input.charAt(i) + " skipped");
                }
                else {
                    badLetters.add(input.charAt(i));
                    System.out.println("Added bad letter: " + input.charAt(i));
                }

            }
            else if (result.charAt(i) == 'Y' || result.charAt(i) == 'y') {
                if (possibleLetters.contains(input.charAt(i))){
                    System.out.println("Letter " + input.charAt(i) + " skipped");
                }
                else {
                    possibleLetters.add(input.charAt(i));
                    System.out.println("Added possible letter: " + input.charAt(i));
                }
            }
            else{
                finWord[i] = input.charAt(i);
                //finalWord.add(i, input.charAt(i));
                dict.put(input.charAt(i),i);//key = char, value = index
                System.out.println("Added letter " + input.charAt(i) + " to the final word at index " + i);

            }
        }
        System.out.println("\nFinished looping through the results...");
        System.out.println("Successfully added all good or bad letters");
        printStatus();
    }



    public void wordRemover() {
        for (int i = 0; i < badLetters.size(); i++){ //for each character in the array
            for (int j = wordList.size() - 1 ; j >= 0; j--) { //for each word on the word list
                String s  = wordList.get(j);
                for (int k = 0; k < s.length(); k++) { //for each words letter
                    if (s.charAt(k) == badLetters.get(i)){
                        System.out.println("Removed word '" + wordList.get(j)+ "' from the list because the letter " + s.charAt(k) + " is at index " + k + " and is equal to badLetter " + badLetters.get(i));
                        wordList.remove(j);
                        break;
                    }
                }
            }
        }

        for (Map.Entry <Character, Integer> entry : dict.entrySet()){
            System.out.println("Map remover is entered 1");
            for (int a = wordList.size() - 1 ; a >= 0; a--){
                String s  = wordList.get(a);
                if (s.charAt(entry.getValue()) != entry.getKey()){
                    System.out.println("Map remover is entered 2, removed word " + wordList.get(a) + " because " + s.charAt(entry.getValue()) + " doesnt equal " + entry.getKey());
                    wordList.remove(a);
                }
                        /*for (int b = 0; b < s.length(); b++){ //for each words letter
                            if (s.charAt(b) != entry.getKey()){
                                wordList.remove(a);
                                break;
                            }
                        }*/
            }
        }

        printPossibleWords();

    }
}

public class wordleSolver{

    public static void main (String[] args){
        String line;
        int i = 0;

        List<String> wordList = new ArrayList<String>();

        try (BufferedReader br = new BufferedReader(new FileReader("sgb-words.txt"))) {
            while((line= br.readLine()) != null){
                wordList.add(line);
                //System.out.println(wordList.get(i++)); //used to test if the words are correctly inputted into the array
            }
            wordleChecker wordle1 = new wordleChecker(wordList);
            wordle1.initCheck();


        }

        catch (Exception e){

        }
    }

}

File: chromeext/wordlesolver.js
const fs = require('fs');

class WordleChecker {
  constructor(wordList) {
    this.wordList = wordList;
    this.badWordList = [];
    this.running = true;

    this.dict = new Map();
    this.finWord = new Array(5);
    this.word = '';
    this.result = '';
    this.possibleLetters = [];
    this.badLetters = [];
    this.finalWord = [];
    this.allGreen = false;
  }

  initCheck() {
    const stdin = require('readline-sync');
    console.log("What word was entered? (Common starter word is Crane)");
    this.word = stdin.question();

    console.log("Successfully initialized word: " + this.word);

    console.log("Enter the results of the initial guess in order from first letter to last letter (B = black (not used), G = green (correct spot), Y = yellow (used, but in the wrong spot), for example: BGYYB");
    this.result = stdin.question();
    this.letterSorter(this.word, this.result);

    this.wordRemover();
    console.log("wordRemover() finished");

    while (this.running) {
      this.checker();
    }
  }

  checker() {
    const stdin = require('readline-sync');
    console.log("What is the next word entered?");
    this.word = stdin.question();
    console.log("Successfully initialized word: " + this.word);
    console.log("Enter the results of the initial guess in order from first letter to last letter (B = black (not used), G = green (correct spot), Y = yellow (used, but in the wrong spot), for example: BGYYB");
    this.result = stdin.question();
    this.letterSorter(this.word, this.result);
    this.wordRemover();
    console.log("Is the word finished? Y/N");
    const yn = stdin.question();
    if (yn === "Y" || yn === "y") {
      this.running = false;
    }
  }

  printStatus() {
    console.log("The possible letters so far are: ");
    for (let i = 0; i < this.possibleLetters.length; i++) {
      process.stdout.write(this.possibleLetters[i] + " ");
    }
    console.log("");

    console.log("The bad letters so far are: ");
    for (let i = 0; i < this.badLetters.length; i++) {
      process.stdout.write(this.badLetters[i] + " ");
    }
    console.log("");

    console.log("The word we have so far is: ");
    for (let i = 0; i < this.finWord.length; i++) {
      process.stdout.write(String(this.finWord[i]));
    }
    console.log();
  }

  printPossibleWords() {
    for (let i = 0; i < this.wordList.length; i++) {
      if (i % 10 === 0) {
        process.stdout.write("\n");
      }
      process.stdout.write(this.wordList[i] + " ,");
    }
    console.log("There are " + this.wordList.length + " words");
  }

  letterSorter(input, result) {
    for (let i = 0; i < result.length; i++) {
      if (result.charAt(i) === 'B' || result.charAt(i) === 'b') {
        if (this.badLetters.includes(input.charAt(i)) || this.possibleLetters.includes(input.charAt(i))) {
          console.log("Letter " + input.charAt(i) + " skipped");
        } else {
          this.badLetters.push(input.charAt(i));
          console.log("Added bad letter: " + input.charAt(i));
        }
      } else if (result.charAt(i) === 'Y' || result.charAt(i) === 'y') {
        if (this.possibleLetters.includes(input.charAt(i))) {
          console.log("Letter " + input.charAt(i) + " skipped");
        } else {
          this.possibleLetters.push(input.charAt(i));
          console.log("Added possible letter: " + input.charAt(i));
        }
      } else {
        this.finWord[i] = input.charAt(i);
        this.dict.set(input.charAt(i), i);
        console.log("Added letter " + input.charAt(i) + " to the final word at index " + i);
      }
    }
    console.log("\nFinished looping through the results...");
    console.log("Successfully added all good or bad letters");
    this.printStatus();
  }

  wordRemover() {
    for (let i = 0; i < this.badLetters.length; i++) {
      for (let j = this.wordList.length - 1; j >= 0; j--) {
        const s = this.wordList[j];
        for (let k = 0; k < s.length; k++) {
          if (s.charAt(k) === this.badLetters[i]) {
            console.log("Removed word '" + this.wordList[j] + "' from the list because the letter " + s.charAt(k) + " is at index " + k + " and is equal to badLetter " + this.badLetters[i]);
            this.wordList.splice(j, 1);
            break;
          }
        }
      }
    }

    for (const [key, value] of this.dict.entries()) {
      for (let a = this.wordList.length - 1; a >= 0; a--) {
        const s = this.wordList[a];
        if (s.charAt(value) !== key) {
          console.log("Map remover is entered 2, removed word " + this.wordList[a] + " because " + s.charAt(value) + " doesnt equal " + key);
          this.wordList.splice(a, 1);
        }
      }
    }

    this.printPossibleWords();
  }
}

// Main function
(async () => {
  const wordList = fs.readFileSync('sgb-words.txt', 'utf8').split('\n');
  const wordle1 = new WordleChecker(wordList);
  wordle1.initCheck();
})();
